src/common/src/util/value_encoding/mod.rs

```diff
@@ -15,94 +15,204 @@
+    let null_tag = data.get_u8();
+    match null_tag {
+        0 => {
+            return Ok(None);
+        }
+        1 => {}
+        _ => {
+            return Err(RwError::from(ErrorCode::InternalError(format!(
+                "Invalid null tag: {}",
+                null_tag
+            ))));
+        }
     }
+    Ok(deserialize_datum(ty, data)?)
 }
 
 /// Deserialize cell bytes which cannot be null into datum.
-pub fn deserialize_cell_not_null(
-    deserializer: &mut value_encoding::Deserializer<impl Buf>,
-    ty: DataType,
-) -> Result<Datum> {
-    match ty {
-        DataType::Decimal => deserialize_decimal(deserializer),
-        _ => Ok(deserialize_datum_not_null_from(
-            ty,
-            deserializer.memcom_de(),
-        )?),
+pub fn deserialize_cell_not_null<B: Buf>(data: &mut B, ty: &DataType) -> Result<Datum> {
+    Ok(deserialize_datum(ty, data)?)
+}
+
+fn serialize_datum(value: ScalarRefImpl, buf: &mut Vec<u8>) {
+    match value {
+        ScalarRefImpl::Int16(v) => buf.put_i16_le(v),
+        ScalarRefImpl::Int32(v) => buf.put_i32_le(v),
+        ScalarRefImpl::Int64(v) => buf.put_i64_le(v),
+        ScalarRefImpl::Float32(v) => buf.put_f32_le(v.into_inner()),
+        ScalarRefImpl::Float64(v) => buf.put_f64_le(v.into_inner()),
+        ScalarRefImpl::Utf8(v) => serialize_str(v.as_bytes(), buf),
+        ScalarRefImpl::Bool(v) => buf.put_u8(v as u8),
+        ScalarRefImpl::Decimal(v) => serialize_decimal(&v, buf),
+        ScalarRefImpl::Interval(v) => serialize_interval(&v, buf),
+        ScalarRefImpl::NaiveDate(v) => serialize_naivedate(v.0.num_days_from_ce(), buf),
+        ScalarRefImpl::NaiveDateTime(v) => {
+            serialize_naivedatetime(v.0.timestamp(), v.0.timestamp_subsec_nanos(), buf)
+        }
+        ScalarRefImpl::NaiveTime(v) => {
+            serialize_naivetime(v.0.num_seconds_from_midnight(), v.0.nanosecond(), buf)
+        }
+        _ => {
+            panic!("Type is unable to be serialized.")
+        }
     }
 }
 
-fn serialize_decimal(decimal: &Decimal) -> Result<Vec<u8>> {
+fn serialize_str(bytes: &[u8], buf: &mut Vec<u8>) {
+    buf.put_u32_le(bytes.len() as u32);
+    buf.put_slice(bytes);
+}
+
+fn serialize_interval(interval: &IntervalUnit, buf: &mut Vec<u8>) {
+    buf.put_i32_le(interval.get_months());
+    buf.put_i32_le(interval.get_days());
+    buf.put_i64_le(interval.get_ms());
+}
+
+fn serialize_naivedate(days: i32, buf: &mut Vec<u8>) {
+    buf.put_i32_le(days);
+}
+
+fn serialize_naivedatetime(secs: i64, nsecs: u32, buf: &mut Vec<u8>) {
+    buf.put_i64_le(secs);
+    buf.put_u32_le(nsecs);
+}
+
+fn serialize_naivetime(secs: u32, nano: u32, buf: &mut Vec<u8>) {
+    buf.put_u32_le(secs);
+    buf.put_u32_le(nano);
+}
+
+fn deserialize_datum<B: Buf>(ty: &DataType, data: &mut B) -> memcomparable::Result<Datum> {
+    Ok(Some(match *ty {
+        DataType::Int16 => ScalarImpl::Int16(data.get_i16_le()),
```

Comment:
The type mapping appears multiple times across the project... Maybe we can unify it by using `serde` in the future (there's a `ScalarImpl::deserialize` accepting a deserializer), but I don't want to intro such framework in value encoding.